apis/v1alpha2/httproute_types.go

```diff
@@ -70,20 +70,30 @@ type HTTPRouteSpec struct {
 	// HTTPRoute rules. If no hostname is specified, traffic is routed
 	// based on the HTTPRouteRules.
 	//
-	// Hostname can be "precise" which is a domain name without the terminating
-	// dot of a network host (e.g. "foo.example.com") or "wildcard", which is
-	// a domain name prefixed with a single wildcard label (e.g. `*.example.com`).
-	// The wildcard character `*` must appear by itself as the first DNS
-	// label and matches only a single label.
-	// You cannot have a wildcard label by itself (e.g. Host == `*`).
 	// Requests will be matched against the Host field in the following order:
 	//
-	// 1. If Host is precise, the request matches this rule if
-	//    the HTTP Host header is equal to Host.
-	// 2. If Host is a wildcard, then the request matches this rule if
+	// 1. If Hostname is precise, the request matches this rule if
+	//    the HTTP Host header is equal to the Hostname.
+	// 2. If Hostname is a wildcard, then the request matches this rule if
 	//    the HTTP Host header is to equal to the suffix
 	//    (removing the first label) of the wildcard rule.
 	//
+	// If a hostname is specified by the Listener that the HTTPRoute is bound
+	// to, all hostnames specified here must match specified hostname as per
+	// the rules above. If any hostname does not match, then the entire Route
+	// is not admitted, and the implementation must raise an 'Admitted'
+	// Condition with a status of `false` for that Listener.
```

Comment:
> the entire Route

This goes against our guidelines: https://gateway-api.sigs.k8s.io/concepts/guidelines/
Specifically these two:
- Prefer not to break things that are working.
- Drop as little traffic as possible.

Could we use the following guideline?
> Make it clear which path has been chosen when a conflict has been identified. Where possible, this should be communicated by setting appropriate status conditions on relevant resources.

So, set the status of the route to reflect that one of the hostname is not admitted but others are?

Reply:
We totally can do that, but the problem is then, how do we communicate that? Right now, we only have a single Condition, `Admitted` to cover all the errors in the Route, so I was using that.

I'd guess in this case, we should add another, negative-polarity (so it's only `status: true` if there's a problem) Condition that covers either the more general "Some config was ignored" case or the the specific "This hostname was ignored" case. 

If we add any Conditions though, we probably should add other Conditions to cover other partial error cases too, I think.

Reply:
So, to be clear, if we're going to make this more nuanced (which I think is a good idea), then we will need more Condition coverage, as `Admitted` as a boolean isn't capable of conveying "mostly okay, some problems".

The two options are:
- add a single new Condition called something like "PartialProblem", or - probably better - "Warning", which would normally be `status: false`, and would change to `status: true` in the case of some error that's not bad enough to mean you need to throw away the whole object (which `Admitted, status: false` implies). This doesn't seem to be how other objects have used Conditions though.
- add extra Conditions for each type of partial thing that could go wrong. In this case, we'd add an `InvalidHostname` Condition that can be `status: true` or `status: false`, without necessarily impacting the `Admitted` state. I like this one better because it's more like the `Node` conditions, which feel like the best use of Conditions to me.

I'll leave this for feedback from @hbagdi, @robscott, @jpeach, or other interested parties. It could also maybe use a document/more detailed discussion, I suppose.

Reply:
I've always thought of hostnames here as being a matching mechanism more than a binding mechanism. Although those concepts are awfully similar, I think there's a distinction here. I've also thought that the hostnames list was interpreted with an OR, not AND.

Let's say a Route owner wants to match requests to foo.com and bar.com, if a Gateway is bound that only supports foo.com I don't think that's an error state. Imagine a separate Gateway binds to this route and supports bar.com. We don't want each set of Gateway conditions in the Route status to include warning messages that they each only support 1 of the 2 hostnames specified. Instead I think it would be more appropriate if they were able to indicate in status which hostnames they did support. 

If there are no Gateway listeners with foo.com or bar.com that select this route, I think that's also fine and more of a lack of selection than an error.

Reply:
There may be O(10k)+ hostnames, I am not sure this is feasible.

Reply:
> I've always thought of hostnames here as being a matching mechanism more than a binding mechanism. 

Yep, I've described it like that before as well. Operationally, an org may end up encoding multiple hostnames into the Route, but only exposing a subset of those through the Gateway Listener. That's alright IMHO. I also think it's OK to not have conditions expressing things that don't happen (in this case, a potential route match that can't happen).

Reply:
For other cases where we have a lot of possible options, we already have an upper limit of 100 in some places, maybe we put that on a `matchedHostnames` field?

If we don't have *something* that tells you what hostnames your Route is selecting though, then a person will need to look at the Gateway and listener hostnames, and take a guess. That does not seem like a good user experience to me.

Reply:
Oh I was thinking this was on the gateway, so aggregating all routes together. On the route it's *likely* less problematic. A route could select 100s of gateways but that seems pretty odd; if we set a limit it would probably handle most cases.

Are there other examples of status fields that are capped and thus are somewhat "best effort"?

Reply:
> Are there other examples of status fields that are capped and thus are somewhat "best effort"?

Within Gateway API, we already limit the number of gateways in Route status to 100: https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha1/shared_types.go#L238. Given the large limit there, we may want to have a smaller limit for number of matched hostnames per GW. 

I agree that a concept like `matchedHostnames` could be quite helpful here in many cases. I do have some questions though:

1. Does a matchedHostname simply represent a hostname from any listener in the GW matching the route?
2. Would this still be reasonable/sensible to compute if Listener was expanded to support multiple hostnames?
3. If the GW has at least one wildcard listener would the matchedHostname list be a direct copy of the route hostname list?
4. What happens if the route hostname list is also empty/unspecified?

Reply:
Let's not get ahead of ourselves here. Is there user experience that suggests we need to highlight this in the API? Do we really need to add more fields or just make the documentation clear to set the expectation?

Reply:
> Let's not get ahead of ourselves here. Is there user experience that suggests we need to highlight this in the API? Do we really need to add more fields or just make the documentation clear to set the expectation?

I do think that this kind of field would be helpful, but yeah, this potential field in status is likely a large enough topic that it could move to a separate issue. No need for us to bundle a new status field with this PR. 

I think we're at a point where we agree on how this should be interpreted and the main question is around how we can make that easier for users to understand. The better documentation that will be added by this PR is a great step and a status field could also be helpful as a follow up.

Reply:
That seems fair. So, how to proceed with Harry's point in the beginning? Right now, with the existing status structure, I don't think there's any way to communicate partial validity. So we have two options:
- we keep what I've put here, and if there's a hostname problem, we throw the whole route out. It seems like noone wants that.
- We allow partial validity, and don't tell people when their config is partially invalid? Or do we put that as a Reason when `Admitted` is `status: true`?

Reply:
I don't think this is partial validity, more partial matching. For example, a Gateway may be listening on "*" but only have DNS pointed to it for foo.example.com. If a Route accepts requests for both foo.example.com and bar.example.com, I don't think we need to add any kind of special status message. I think the real thing we care about here is if there is _no way_ that a request to a Gateway will be routed with the specific Route resource. In that case, I think `Admitted` should be set to false. That includes if there's no hostname intersection between route and listener. I don't think we need to be concerned about a Route accepting requests from other hostnames though. 

I think it's likely going to be relatively common for a Route to accept multiple hostnames, and for those hostnames to be served by different Gateways. I don't think we need any kind of warning state for that.

Reply:
That makes sense to me. I'll try and update this with something to that effect.

Reply:
Okay, this is updated now, I think this one is done?